### PR TITLE
Add a `browser-utils` crate

### PR DESCRIPTION
**Pull request by expenses**
There is a lot of shared logic between the test substrate/polkadot in-browser light clients I've been working on, so I'm putting it in this `browser-utils` crate.

For clarity: I'm still working out how to best approach certain changes needed to get the light clients to work properly, so I'm merging things piece-by-piece.

### CODE REVIEW

File: utils/browser/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+use std::task::Poll;
+use std::pin::Pin;
+use chain_spec::Extension;
+
+pub use libp2p::wasm_ext::{ExtTransport, ffi::Transport};
+
+/// Set up a panic hook and console logging.
+pub fn set_hooks(log_level: log::Level) {
+    console_error_panic_hook::set_once();
+    let _ = console_log::init_with_level(log_level);
```
Comment (expenses) [inline utils/browser/src/lib.rs]: It might be a good idea to re-export these functions instead.

File: bin/node/cli/src/browser.rs [hunk 2]
```diff
@@ -15,45 +15,27 @@
...
 	start_inner(wasm_ext)
+		.await
 		.map_err(|err| JsValue::from_str(&err.to_string()))
 }
 
-fn start_inner(wasm_ext: wasm_ext::ffi::Transport) -> Result<Client, Box<dyn std::error::Error>> {
-	console_error_panic_hook::set_once();
-	console_log::init_with_level(log::Level::Info);
+async fn start_inner(wasm_ext: Transport) -> Result<Client, Box<dyn std::error::Error>> {
+	set_hooks(log::Level::Debug);
```
Comment (gnunicorn) [inline bin/node/cli/src/browser.rs]: Are you intentionally setting a different log level then we had before?

File: utils/browser/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+futures01 = { package = "futures", version = "0.1.29" }
+log = "0.4.8"
+libp2p = { version = "0.13.1", default-features = false }
+console_error_panic_hook = "0.1.1"
+console_log = "0.1.2"
+js-sys = "0.3.22"
+wasm-bindgen = "0.2.55"
+wasm-bindgen-futures = "0.4.5"
+kvdb-web = "0.1.1"
+service = { package = "sc-service", path = "../../client/service", default-features = false }
```
Comment (gnunicorn) [inline utils/browser/Cargo.toml]: will need a version now. The following two lines, too.
```suggestion
service = { version = "2.0.0", package = "sc-service", path = "../../client/service", default-features = false }
```

File: utils/browser/src/lib.rs (hunk 1 above)
Comment (gnunicorn) [reply utils/browser/src/lib.rs]: Yeah, not sure this makes a lot of sense to me. All this does is encapsulate two functions into one and swallow an error. First wanted to suggest you return the error ... but then the difference is almost non existent so I wonder if this function is that useful...

File: bin/node/cli/src/browser.rs (hunk 2 above)
Comment (expenses) [reply bin/node/cli/src/browser.rs]: No, that was an experimental change and I forgot to change it back

File: bin/node/cli/browser-demo/build.sh [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env sh
 wasm-pack build --target web --out-dir ./browser-demo/pkg --no-typescript --release ./.. -- --no-default-features --features "browser"
-python -m SimpleHTTPServer 8000
+python -m http.server 8000
```
Comment (tomaka) [inline bin/node/cli/browser-demo/build.sh]: :+1: Python2 is now deprecated :sweat_smile:

**Comment by tomaka** [comment]
IMO we should be more explicit in the `keystore`. The `keystore` should have a "memory-only" mode rather than using `#[cfg]` stuff. Once the `keystore` can be memory-only, we also no longer need to pass a dummy keystore path in the configuration.

What this PR does is okay-ish but doesn't go in the direction of having clean code.

We can either merge now and fix later, or improve this PR itself.

**Comment by expenses** [comment]
@tomaka What can be cleaned up?

**Comment by tomaka** [comment]
What I mentioned: not having platform-specific code in the `keystore`.

File: client/service/src/builder.rs [hunk 5]
```diff
@@ -250,11 +250,15 @@ where TGen: RuntimeGenesis, TCSExt: Extension {
 		(),
 		TLightBackend<TBl>,
 	>, Error> {
+		#[cfg(not(target_os = "unknown"))]
```
Comment (tomaka) [inline client/service/src/builder.rs]: Well, the next obvious step is to adjust `config.keystore_path` to be an enum of either a path or "in memory".

Comment (expenses) [reply client/service/src/builder.rs]: 👍

**Comment by tomaka** [comment]
This needs a Polkadot PR because of the change in the configuration now though.

**Comment by gavofyork** [comment]
tests not passing either

**Comment by bkchr** [comment]
Needs resolving.

**Comment by expenses** [comment]
Not sure what's causing `wasm-bindgen-webidl` to not compile - that version seems to work fine on my machine and the build server.

**Pull request merged**